CONTRIBUTING.md

```diff
@@ -0,0 +1,572 @@
+ * [CMake coding style](#cmake-coding-style)
+ 
+[Helper Scripts](#helper-scripts)
+ * [Create a new algorithm](#create-a-new-algorithm)
+ * [Converting CUDA code to HIP code](#converting-cuda-code-to-hip-code)
+
+[Writing Tests](#writing-tests)
+ * [Testing know-how](#testing-know-how)
+ * [Some general rules](#some-general-rules)
+ * [Writing tests for kernels](#writing-tests-for-kernels)
+
+[Documentation style](#documentation-style)
+ * [Developer targeted notes](#developer-targeted-notes)
+ * [Whitespaces](#whitespaces)
+ * [Documenting examples](#documenting-examples)
+ 
+[Other general programming comments](#other-general-programming-comments)
+ * [C++ standard stream objects](#c++-standard-stream-objects)
+ * [Warnings](#warnings)
+ * [Avoiding circular dependencies](#avoiding-circular-dependencies)
+
+
+## Most important stuff (A TL;DR)
+
+* `GINKGO_DEVEL_TOOLS` needs to be set to `on` to commit. This requires
+  `clang-format` to be installed. See [Automatic code
+  formatting](#automatic-code-formatting) for more details. Once installed, you
+  can run `make format` in your `build/` folder to automatically format your
+  modified files. Once formatted, you can commit your changes.
+
+* See [Our git workflow](#our-git-workflow) to get a quick overview of our
+  workflow.
+
+* See [Creating, Reviewing and Merging Pull
+  Requests](#creating-reviewing-and-merging-pull-requests) on how to create a
+  Pull request.
+
+
+## Project structure
+
+Ginkgo is divided into a `core` module with common functionalities independent
+of the architecture, and several kernel modules (`reference`, `omp`, `cuda`,
+`hip`) which contain low-level computational routines for each supported
+architecture.
+
+### Extended header files
+
+Some header files from the core module have to be extended to include special
+functionality for specific architectures. An example of this is
+`core/base/math.hpp`, which has a GPU counterpart in `cuda/base/math.hpp`. For
+such files you should always include the version from the module you are working
+on, and this file will internally include its `core` counterpart.
+
+### Using library classes
+
+You can use and call functions of existing classes inside a kernel (that are
+defined and not just declared in a header file), however, you are not allowed to
+create new instances of a polymorphic class inside a kernel (or in general
+inside any kernel module like cuda/hip/omp/reference) as this creates circular
+dependencies between the `core` and the backend library. With this in mind, our
+CI contains a job which checks if such a circular dependency exists.
+
+For example, when creating a new matrix class `AB` by combining existing classes
+`A` and `B`, the `AB::apply()` function composed of kernel invocations to
+`A::apply()` and `B::apply()` can only be defined in the core module, it is not
+possible to create instances of `A` and `B` inside the `AB` kernel file. An
+example for such a class the `hybrid` matrix format.
```

Comment:
As the previous place to comment seems to have disappeared after the format and rebase, I start the discussion here again.

@thoasm , @tcojean : how about this ?
```suggestion
For example, when creating a new matrix class `AB` by combining existing classes
`A` and `B`, the `AB::apply()` function composed of invocations to
`A::apply()` and `B::apply()` can only be defined in the core module, it is not
possible to create instances of `A` and `B` inside the `AB` kernel files. This is 
to avoid the aforementioned circular dependency issue. An example for such
 a class the `Hybrid` matrix format, which uses the `apply()` of the `Ell` and `Coo`
 matrix formats. Nevertheless, it is possible to call the kernels themselves directly 
 within the same executor. For example, `cuda::dense::add_scaled()` can be called from
 any other `cuda` kernel. 
```

Reply:
I'm fine with this.

Reply:
Yes, this looks good to me, well done.